core/src/Kokkos_LogicalSpaces.hpp

```diff
@@ -0,0 +1,433 @@
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_LOGICALSPACES_HPP
+#define KOKKOS_LOGICALSPACES_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Core.hpp>
+#include <Kokkos_Core_fwd.hpp>
+#include <Kokkos_ScratchSpace.hpp>
+#include <impl/Kokkos_MemorySpace.hpp>
+#include <impl/Kokkos_Error.hpp>
+#include <impl/Kokkos_SharedAlloc.hpp>
+#include <impl/Kokkos_Profiling_Interface.hpp>
+#include <cstring>
+namespace Kokkos {
+namespace Experimental {
+struct DefaultMemorySpaceNamer {
+  static constexpr const char* get_name() {
+    return "DefaultLogicalMemorySpaceName";
+  }
+};
+
+/// \class LogicalMemorySpace
+/// \brief
+///
+/// LogicalMemorySpace is a space that is identical to another space,
+/// but differentiable by name and template argument
+
+template <class BaseSpace, class DefaultExecutionSpace = void,
+          class Namer               = DefaultMemorySpaceNamer,
+          bool SharesAccessWithBase = true>
+class LogicalMemorySpace {
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+  // [DZP] For some reason I don't yet know, using LogicalMemorySpaces
+  // inside an OpenMPTarget build causes errors in the
+  // SharedAllocationRecords of other types. This is my way of erorring
+  // a build if we instantiate a LogicalMemSpace in an OMPTarget build
+  static_assert(!std::is_same<BaseSpace, BaseSpace>::value,
+                "Can't use LogicalMemorySpaces in an OpenMPTarget build, we're "
+                "debugging memory issues");
+#endif
+ public:
+  //! Tag this class as a kokkos memory space
+  using memory_space = LogicalMemorySpace<BaseSpace, DefaultExecutionSpace,
+                                          Namer, SharesAccessWithBase>;
+  using size_type    = size_t;
+
+  /// \typedef execution_space
+  /// \brief Default execution space for this memory space.
+  ///
+  /// Every memory space has a default execution space.  This is
+  /// useful for things like initializing a View (which happens in
+  /// parallel using the View's default execution space).
+
+  using execution_space =
+      typename std::conditional<std::is_void<DefaultExecutionSpace>::value,
+                                typename BaseSpace::execution_space,
+                                DefaultExecutionSpace>::type;
+
+  using device_type = Kokkos::Device<execution_space, memory_space>;
+
+  BaseSpace underlying_allocator;
```

Comment:
Why is this public? (blocking)

Also probably not the best name since it is a memory space not an allocator.  (not blocking)

Reply:
"Carelessness" and "true," respectively. Will fix both